src/core/lib/event_engine/ares_resolver.cc

```diff
@@ -0,0 +1,714 @@
+
+#include <stdint.h>
+
+#include <string>
+#include <vector>
+
+#include "src/core/lib/iomgr/port.h"
+
+// IWYU pragma: no_include <arpa/inet.h>
+// IWYU pragma: no_include <arpa/nameser.h>
+// IWYU pragma: no_include <inttypes.h>
+// IWYU pragma: no_include <netdb.h>
+// IWYU pragma: no_include <netinet/in.h>
+// IWYU pragma: no_include <stdlib.h>
+// IWYU pragma: no_include <sys/socket.h>
+// IWYU pragma: no_include <ratio>
+
+#if GRPC_ARES == 1
+
+#include <ares_nameser.h>
+#include <string.h>
+
+#include <algorithm>
+#include <chrono>
+#include <initializer_list>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+#include "absl/functional/any_invocable.h"
+#include "absl/hash/hash.h"
+#include "absl/strings/match.h"
+#include "absl/strings/numbers.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_format.h"
+#include "absl/types/optional.h"
+
+#include <grpc/event_engine/event_engine.h>
+#include <grpc/support/log.h>
+
+#include "src/core/lib/address_utils/parse_address.h"
+#include "src/core/lib/address_utils/sockaddr_utils.h"
+#include "src/core/lib/event_engine/grpc_polled_fd.h"
+#include "src/core/lib/event_engine/time_util.h"
+#include "src/core/lib/gprpp/debug_location.h"
+#include "src/core/lib/gprpp/host_port.h"
+#include "src/core/lib/gprpp/ref_counted_ptr.h"
+#include "src/core/lib/iomgr/resolved_address.h"
+#include "src/core/lib/iomgr/sockaddr.h"
+#ifdef GRPC_POSIX_SOCKET_ARES_EV_DRIVER
+#include "src/core/lib/event_engine/posix_engine/tcp_socket_utils.h"
+#elif defined(GRPC_WINDOWS_SOCKET_ARES_EV_DRIVER)
+#endif
+
+namespace grpc_event_engine {
+namespace experimental {
+
+grpc_core::TraceFlag grpc_trace_ares_resolver(false,
+                                              "event_engine_ares_resolver");
+
+namespace {
+
+absl::Status AresStatusToAbslStatus(int status, absl::string_view error_msg) {
+  switch (status) {
+    case ARES_ECANCELLED:
+      return absl::CancelledError(error_msg);
+    case ARES_ENOTIMP:
+      return absl::UnimplementedError(error_msg);
+    case ARES_ENOTFOUND:
+      return absl::NotFoundError(error_msg);
+    default:
+      return absl::UnknownError(error_msg);
+  }
+}
+
+EventEngine::Duration calculate_next_ares_backup_poll_alarm_duration() {
+  // An alternative here could be to use ares_timeout to try to be more
+  // accurate, but that would require using "struct timeval"'s, which just
+  // makes things a bit more complicated. So just poll every second, as
+  // suggested by the c-ares code comments.
+  return std::chrono::seconds(1);
+}
+
+bool IsIpv6LoopbackAvailable() {
+#ifdef GRPC_POSIX_SOCKET_ARES_EV_DRIVER
+  return PosixSocketWrapper::IsIpv6LoopbackAvailable();
+#elif defined(GRPC_WINDOWS_SOCKET_ARES_EV_DRIVER)
+  // TODO(yijiem): make this portable for Windows
+  return true;
+#else
+#error "Unsupported platform"
+#endif
+}
+
+absl::Status SetRequestDNSServer(absl::string_view dns_server,
+                                 ares_channel channel) {
```

Comment:
Done. `ares_channel` is actually a pointer type so that's why things still work. But I have changed here and elsewhere to pass in as a pointer to make the semantic clear.